src/Algebra/Graph.hs

```diff
@@ -173,12 +173,16 @@ compatible with 'overlay' and 'connect' operations:
 @'empty' <= x
 x     <= x + y
 x + y <= x * y@
+
+Deforestation (fusion) is implemented for some functions in this module. This means
+that when a function tagged as a \"good producer\" is composed with a \"good consumer\",
+the intermediate structure will not be built.
 -}
 data Graph a = Empty
              | Vertex a
              | Overlay (Graph a) (Graph a)
              | Connect (Graph a) (Graph a)
-             deriving (Show, Generic)
+             deriving (Generic)
```

Comment:
I think it's best to continue using derived `Show`. Performance is not really that important for `Show` because it is not supposed to be used for large graphs anyway. Using derived `Show` allows us to avoid the extra complexity that comes with the custom implementation.